### PR TITLE
Stop in progress QC voting when network goes into EFM

### PR DESCRIPTION
**Pull request by kc1116**
This PR updates the epoch manager engine to process EpochEmergencyFallbackTriggered events and exit QC Voting retries when those events are processed.

ref: https://github.com/onflow/flow-go/issues/5733

### CODE REVIEW

File: module/epochs/qc_voter.go [hunk 1]
```diff
@@ -170,13 +172,29 @@ func (voter *RootQCVoter) Vote(ctx context.Context, epoch protocol.Epoch) error
 		// update our last successful client index for future calls
 		voter.updateLastSuccessfulClient(clientIndex)
 		return nil
+	}
+	err = retry.Do(ctx, backoff, func(ctx context.Context) error {
```
Comment (AlexHentschel) [inline module/epochs/qc_voter.go:176]: have you considered restructuring the code such that we can abort the voting by cancelling the context?

**Review by AlexHentschel — CHANGES_REQUESTED** [review]
I think the implementation approach in this PR has conceptual problems in some very critical aspects:
1. I believe voting cannot be cancelled before the voting process itself already completed, which defeats the point. (see [my comment here]())
2. `StopVoting` call must happen while the voting is still ongoing. Otherwise, the current code breaks the future voting cycle, which likely is for a valid new epoch where we _must_ participate. In its current form, I think a single EFM would lead to all collectors abstaining from voting for their respective root QC.
3. Already previously, the `epochmgr.Engine` was not quite compliant with "Flow core protocol being an information-driven system" (see [talk here](https://drive.google.com/file/d/10QncGnc5zotKUq2vHgDt1hdYdSIMllex/view?usp=drive_link) for further details). The reason is that we use channels with capacity 1: https://github.com/onflow/flow-go/blob/b9295e00aa2fc2899a38bd2932b87eb070a6873a/engine/collection/epochmgr/engine.go#L92-L95 if a notification (e.g. `EpochSetupPhaseStarted`) is repeated, it will block the code emitting the notification.

   Previously, the lack of Idempotency (ability to make repeated calls without adverse effects) was not great, but it would only slow down the node. With the revisions now, we are adding correctness problems, because we are potentially undermining _future_ voting rounds.

## Desired behaviour
* all notifications and API-level method calls are [idempotent](https://en.wikipedia.org/wiki/Idempotence) ideally
* when we observe an `EpochEmergencyFallbackTriggered` notification, we cancel the voting process _currently_ ongoing. It is a no-op if there currently is no voting. 
  * Note that golang channels do not provide [`Strict Serializability`](http://www.bailis.org/blog/linearizability-versus-serializability/#:~:text=serializability%3A%20multi-operation%2C%20multi-object%2C%20arbitrary%20total%20order) (behavior is equivalent to some serial execution, and the serial order corresponds to real time). My gut feeling is that we need this property here. Otherwise, we can't express things like "only cancel if we are currently voting" I think you probably need to resort to atomics/mutex in the `RootQCVoter`, as they have the stronger guarantees
* Be careful with the current guarantee that the `RootQCVoter` provides: ["It is safe to run Vote multiple times within a single setup phase"](https://github.com/onflow/flow-go/blob/b9295e00aa2fc2899a38bd2932b87eb070a6873a/module/epochs.go#L16-L17). My preference would be to maintain this guarantee, but that might be hard. 
  * Once option could be for the `RootQCVoter` to remember which voting processes are currently ongoing. But I think adding that sort of tracking would increase the `RootQCVoter`'s complexity quite a bit. 
  * I [suggested]() to use the context that we anyway input into the `Vote` function as a means to also cancel the voting. Thereby, tracking which votes are currently ongoing could be delegated to the `epochmgr.Engine` ... and in the engine, we can much easier argue that at most one voting process is ongoing at the same time.

File: engine/collection/epochmgr/engine.go
[the remark this answers is not in the document]
Comment (jordanschalm) [reply engine/collection/epochmgr/engine.go:1]: You're right - I'm glad you noticed this. This should be passing in `epochCtx`, not `engineCtx`. 

The `engineCtx` errors are already handled at the node scaffold level, the goal here is to capture errors from epoch-scoped components, so `epochCtx` should be passed.

**Comment by kc1116** [comment]
@AlexHentschel I have restructured the code around cancelling the context

File: engine/collection/epochmgr/engine.go [hunk 2]
```diff
@@ -56,6 +57,9 @@ type Engine struct {
 	mu     sync.RWMutex                       // protects epochs map
 	epochs map[uint64]*RunningEpochComponents // epoch-scoped components per epoch
 
+	qcVoteMu          sync.Mutex                    // protects qc votes map
+	inProgressQCVotes map[string]context.CancelFunc // map of context cancel funcs, each cancel func maps to exactly 1 in progress qc voting process
+
```
Comment (AlexHentschel) [inline engine/collection/epochmgr/engine.go:62]: To the best of my understanding, there can only be one voting process. Therefore, a map and a mutex are unnecessarily complex, I think. Instead, I would suggest to use an atomic

```suggestion
	inProgressQCVote *atomic.Pointer[context.CancelFunc]
```

File: engine/collection/epochmgr/engine.go [hunk 3]
```diff
@@ -537,11 +558,36 @@ func (e *Engine) removeEpoch(counter uint64) {
...
+// stopAllInProgressQcVotes cancels the context for all in progress root qc voting.
+func (e *Engine) stopAllInProgressQcVotes() {
+	e.qcVoteMu.Lock()
+	for processId, cancel := range e.inProgressQCVotes {
+		cancel()
+		delete(e.inProgressQCVotes, processId)
+		e.log.Warn().Msgf("root qc vote process canceled: %s", processId)
+	}
+	e.qcVoteMu.Unlock()
+}
```
Comment (AlexHentschel) [inline engine/collection/epochmgr/engine.go:593]: with my suggestion to use atomics, this function could look like: 
```suggestion
// stopAllInProgressQcVotes cancels the context for the root QC voting (if there is any currently ongoing).
func (e *Engine) stopAllInProgressQcVotes() {
	for {
		c := e.inProgressQCVote.Load()
		if c != nil {
			e.log.Warn().Msgf("voting for cluster root block cancelled")
			(*c)() // cancel
		}
		if e.inProgressQCVote.CompareAndSwap(c, nil) {
			break
		}
	}
}
```

**Review by AlexHentschel — APPROVED** [review]

File: engine/collection/epochmgr/engine.go (hunk 2 above)
Comment (kc1116) [reply engine/collection/epochmgr/engine.go]: https://github.com/onflow/flow-go/pull/5991/commits/091264c110f69ab77396c226cf45b98e9df02ba0

File: module/epochs/qc_voter.go [hunk 4]
```diff
@@ -170,7 +168,8 @@ func (voter *RootQCVoter) Vote(ctx context.Context, epoch protocol.Epoch) error
 		// update our last successful client index for future calls
 		voter.updateLastSuccessfulClient(clientIndex)
 		return nil
-	})
+	}
+	err = retry.Do(ctx, backoff, castVote)
```
[the remark this answers is not in the document]
Comment (kc1116) [reply module/epochs/qc_voter.go:172]: https://github.com/onflow/flow-go/pull/5991/commits/091264c110f69ab77396c226cf45b98e9df02ba0

File: module/epochs/qc_voter_test.go [hunk 5]
```diff
@@ -109,16 +111,46 @@ func (suite *Suite) TestInvalidPhase() {
...
+		suite.Assert().True(epochs.IsClusterQCNoVoteError(err), "got error of unexpected type")
+		close(voteReturned)
+	}()
+
+	select {
+	case <-voteReturned:
+		// expected behaviour
+	case <-time.NewTimer(1000 * time.Second).C:
+		suite.Fail("call of `Vote` method has not returned within the test's timeout")
+	}
```
Comment (jordanschalm) [inline module/epochs/qc_voter_test.go:155]:
```suggestion
	unittest.AssertClosesBefore(suite.T(), voteReturned, time.Second, "call of `Vote` method has not returned within the test's timeout")
```

File: module/epochs/qc_voter_test.go [hunk 6]
```diff
@@ -109,16 +111,46 @@ func (suite *Suite) TestInvalidPhase() {
...
 	err := suite.voter.Vote(context.Background(), suite.epoch)
 	suite.Assert().NoError(err)
 }
+
+// TestCancelVoting verifies correct behaviour when the context injected into the `Vote` method is cancelled.
+// The `RootQCVoter` should abort voting and quickly return an `ClusterQCNoVoteError`.
+func (suite *Suite) TestCancelVoting() {
+	// We emulate the case, where the `QCContractClient` always returns an expected sentinel error, indicating
+	// that it could not interact with the system smart contract. To create a realistic test scenario, we cancel
+	// the context injected into the voter, when it is trying to submit a vote for the first time.
```
Comment (jordanschalm) [inline module/epochs/qc_voter_test.go:133]:
```suggestion
	// the context injected into the voter, when it is trying to submit a vote for the first time.
	// The returned transient error will cause a retry, during which the retry logic will observe the context
	// has been cancelled and exit with the context error. 
```

**Review by jordanschalm — APPROVED** [review]
Added several suggestions for simplifications and documentation changes, otherwise looks good.

File: engine/collection/epochmgr/engine.go [hunk 7]
```diff
@@ -444,14 +460,20 @@ func (e *Engine) prepareToStopEpochComponents(epochCounter, epochMaxHeight uint6
...
 	if err != nil {
 		if epochs.IsClusterQCNoVoteError(err) {
 			e.log.Warn().Err(err).Msg("unable to submit QC vote for next epoch")
 			return
 		}
 		ctx.Throw(fmt.Errorf("unexpected failure to submit QC vote for next epoch: %w", err))
 	}
+
+	e.inProgressQCVote.Swap(nil)
+	cancel()
```
[the remark this answers is not in the document]
Comment (kc1116) [reply engine/collection/epochmgr/engine.go]: https://github.com/onflow/flow-go/pull/5991/commits/f086cc9c0d5e1a89e9847d67d7a6c92f06d78288

File: engine/collection/epochmgr/engine.go [hunk 8]
```diff
@@ -56,6 +58,8 @@ type Engine struct {
 	mu     sync.RWMutex                       // protects epochs map
 	epochs map[uint64]*RunningEpochComponents // epoch-scoped components per epoch
 
+	inProgressQCVote *atomic.Pointer[CancelFuncWrapper] // tracks the cancel callback of the in progress QC vote
```
[the remark this answers is not in the document]
Comment (kc1116) [reply engine/collection/epochmgr/engine.go]: https://github.com/onflow/flow-go/pull/5991/commits/baf45b3e997f525382dec90d2a4e3e32b19bdf04

File: engine/collection/epochmgr/engine_test.go [hunk 9]
```diff
@@ -532,3 +532,53 @@ func (suite *Suite) TestRespondToEpochTransition() {
 	// the expired epoch should have been stopped
 	suite.AssertEpochStopped(suite.counter - 1)
 }
+
+// TestStopQcVoting tests that, if we encounter an EpochEmergencyFallbackTriggered event
+// the engine will stop in progress QC voting. The engine keeps track of current in process
+// qc voting by keeping a map of context cancel funcs. When the EFM event is encountered each
+// cancel func will be invoked and all voting processes will stop.
```
[the remark this answers is not in the document]
Comment (kc1116) [reply engine/collection/epochmgr/engine_test.go]: https://github.com/onflow/flow-go/pull/5991/commits/a3679318c2c66c0a10a08b7a67cf8075f03a95ae

**Pull request merged**